### PR TITLE
[com_fields] Move some events from system plugin to content plugin v2

### PR DESCRIPTION
**Pull request by piotr-cz**
Pull Request for Issue -.

### Summary of Changes
Plugin events are triggered by the dispatcher in order specified by the plugins administration component. However some plugin types like system plugins are dispatched before content plugins.

The pull requests moves _onContentAfterTitle_, _onContentBeforeDisplay_ and _onContentAfterDisplay_ from fields system plugin to fields content plugin - so it's possible to define where output will be rendered on the page.

I've prepared an example where I'm using both vote plugin and fields plugin to show a gallery.  Both are configured to show on _onContentBeforeDisplay_ event.
The goal is to 

### Testing Instructions
1) Switch On the vote plugin
  _Extensions > Plugins > Content - vote_, make sure the _Position_ is set to _Top_
2) Add new Custom field
  _Content > Fields > New_, make sure the _Automatic Display_ is set to _Before Display_
3) Change ordering o plugins 
_Extensions > Plugins >  Search Tools_
Select Type: Content, Sort table by: Ordering Ascending

### Expected result
User can choose the order of content plugins to display
j-fields-after

### Actual result
The fields plugin output is always displayed before any other content plugins.
j-fields-before

### Documentation Changes Required
none

cc @laoneo

### CODE REVIEW

**Comment by laoneo** [comment]
Nice one. Guess it's the only way to use the fields after some other content plugin. Should we not move all onContent events to this plugin? The next move would be then to move the user events to a user field plugin and the finder to also a separate plugin. Line that we can eliminate the system plugin.

**Comment by piotr-cz** [comment]
Probably yes, we should move all the event listeners but let's hear others opinions.
Moreover this is more correct when it comes to syntax.

**Comment by mbabker** [comment]
Honestly, this is a bandaid fix.  We really need to improve the UI management as a whole.  It really should be possible to globally order plugins, not be locked in based on the group they are imported with.

**Comment by piotr-cz** [comment]
@mbabker It's up to another discussion if user should be able to change plugins order outside their types/ groups.
I see some sense in that system plugins are triggered before content plugins.

For now just should make sure that plugins don't register events that are not supposed to - like here system plugin registering content events.

**Comment by mbabker** [comment]
There is nothing saying "X plugin group should not register events in Y plugin group".  I actually think that would be a bad standard to enforce, then again I've also made it clear in the past that I'm not seeing the benefit of this group structure anymore other than to limit the number of loaded event listeners based on an arbitrary code path which results in weird behaviors and inconsistent results.

**Comment by piotr-cz** [comment]
The only and IMHO important benefit I see is that you may easily filter plugins by group in the administration area (this is really helpful as there are around 66 native plugins).

This really helps when you want to see which plugins are manipulating the content, disable all quickicons plugins and so on.

But again each group should register own events otherwise filtering functionality is useless.

**Comment by mbabker** [comment]
And I'd say the ordering functionality is flawed.  You should be able to order plugins globally, not only within a specific filter group.

We should not be continuing to promote the mentality of "system plugins only for `onAfterRender`" or "content plugins only for `onContent*`" type logic.

The filtering logic is good for the UI.  But IMO it doesn't translate as well into the code level.

**Comment by piotr-cz** [comment]
To sum up: to fix the issue I've described we can either implement global ordering of plugins or keep events within groups like in the above commit.

**Comment by mbabker** [comment]
For where we're at now, it's fine.  I'd just like to see us move away from creating an artificial separation of these things and make plugins and their grouping/ordering integrate better at a global scale (so you can order a content plugin before a system plugin for when those have been imported, though I'd still like to phase out importing of different groups at different times but that's probably a battle I'll never win).

**Comment by Bakual** [comment]
The downside of this is that you can't turn of the content plugin if you don't want it to be parsing the article content for possible field tags. You would turn off also the automatic display.
Other than that it should be fine.

I wouldn't extend it to other events. In this case it's fine since the content plugin already exists, but if you have to create a new field plugin for eg user it gets out of hand fast imho.

I also agree with Michael that there is a serious flaw in our plugin system. It would be great if plugins really would "register" the events they want to listen to. Currently it's the dispatcher checking each plugin if there is a matching method specified, which is also why we have the groups. So the dispatcher doesn't have to always look into hundred plugin classes to see what methods they support.

**Comment by piotr-cz** [comment]
@Bakual You are right, I didn't know that there is a functionality of inserting fields using _{field}_ and _{fieldgroup}_ tags.

But still I think that the PR is fine as it is. If somebody won't want to display fields, then (s)he should disable  content plugin and it will be disabled for both cases - for manual tags as well as for automatic display as it refers to content.

**Comment by Bakual** [comment]
>  If somebody won't want to display fields

If someone doesn't want to display fields, he should disable the system - fields plugin because that one actually does the bulk of the work (and performance hit).
With this PR, you would always either have both plugins enabled or disabled. Having one enabled and the other disabled doesn't make any sense.

Just some explanation: The content plugin exists because I created it to support those field tags in articles. It goes together with the fields editor button plugin. This feature was added after fields were already working. It is the sole purpose of that plugin currently.

I get the need for the ordering, and for that I can see the purpose of this PR as a workaround because we currently can't order the plugins across groups.
Imho the better fix would be to allow the ordering of plugins independant of plugin group. But that may be tricky as well, I don't know.

**Comment by laoneo** [comment]
> you would always either have both plugins enabled or disabled

Not necessarily, with the content plugin disabled, you can turn off automatic display only. The rest is still working as expected.

**Comment by Bakual** [comment]
Yeah. You can still add values to the fields and create them. But without template overrides, the values will not show anywhere.
So I think it is at least a rare case where only the content plugin would be disabled.

**Comment by laoneo** [comment]
It's true for the core, but perhaps some 3rd party extensions only want to work on the fields array and don't want to automatic display. I know it is just guessing. But somehow I like the idea to not have a global system plugin which is initialized on every page request, but only plugins which are actually needed.

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 887a0cf84add0d20905c7e5dfa62dc79fb729cff<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by ghost** [comment]
This is an easy Test (having fine Test Instructions) which needs a second Test.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by fancyFranci** [comment]
I have tested this item :white_check_mark: successfully on 887a0cf84add0d20905c7e5dfa62dc79fb729cff<br /><br />The patch sets the voting above the custom field.
But unlike in the instructions, I needed to set the ordering inside the vote plugin and not in the ordering of plugins.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by ghost** [comment]
@piotr-cz can you please resolve conflicting File so this Pull Request can be set "Ready to Commit"?

**Comment by ghost** [comment]
can anyone than @piotr-cz resolve conflicting Files?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by piotr-cz** [comment]
@franz-wohlkoenig ok, done

**Comment by ghost** [comment]
I have tested this item :red_circle: unsuccessfully on 627a401531302179064577cc78135d4bd8b9444d

After applying Patch got in Frontend:

bildschirmfoto 2017-12-28 um 10 59 17

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by ghost** [comment]
If this PR get no Response, it will be closed at 28th February 2018.

**Comment by fancyFranci** [comment]
@piotr-cz please just add the two missing ``use Joomla\CMS\...`` statements and it will work again.

**Comment by piotr-cz** [comment]
@FPerisa which ones?
Sorry, but I'm out of loop for long time

**Comment by ghost** [comment]
Issue stay open.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by fancyFranci** [comment]
@piotr-cz
You need to move all three use statements
```
use Joomla\Registry\Registry;
use Joomla\CMS\Factory;
use Joomla\CMS\Language\Multilanguage;
```
from system\fields\fields.php to content\fields\fields.php (and remove it from the first one).

But coming from another testing, I found a new error. The "/tagged-items" page throws an exception, because you need to copy the system fields `private function prepareTagItem($item)` to the content one too!

But after all this it is finally working :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14875">issues.joomla.org/tracker/joomla-cms/14875</a>.</sub>

**Comment by piotr-cz** [comment]
@FPerisa 
I've moved these dependencies and copied `prepareTagItem` method over to content plugin.

Thanks for assistance.

**Comment by ghost** [comment]
@piotr-cz can you please resolve the conflicting files so this can be tested?

**Comment by sudhir-j-sapkal** [comment]
@franz-wohlkoenig - Is it OK to send separate PR with solving the conflicts , If @piotr-cz is not available ?

**Comment by piotr-cz** [comment]
@sudhir-j-sapkal Please do so

File: plugins/content/fields/fields.php [hunk 1]
```diff
@@ -165,4 +171,175 @@ private function prepare($string, $context, $item)
...
+	 * The display event.
+	 *
+	 * @param   string    $context     The context
+	 * @param   stdClass  $item        The item
+	 * @param   Registry  $params      The params
+	 * @param   integer   $limitstart  The start
+	 *
+	 * @return  string
+	 *
+	 * @since   3.7.0
```
Comment (PhilETaylor) [inline plugins/content/fields/fields.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: plugins/content/fields/fields.php [hunk 2]
```diff
@@ -165,4 +171,175 @@ private function prepare($string, $context, $item)
...
```
Comment (PhilETaylor) [inline plugins/content/fields/fields.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/content/fields/fields.php [hunk 3]
```diff
@@ -165,4 +171,175 @@ private function prepare($string, $context, $item)
...
```
Comment (PhilETaylor) [inline plugins/content/fields/fields.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/content/fields/fields.php [hunk 4]
```diff
@@ -165,4 +171,175 @@ private function prepare($string, $context, $item)
...
+	 * Performs the display event.
+	 *
+	 * @param   string    $context      The context
+	 * @param   stdClass  $item         The item
+	 * @param   Registry  $params       The params
+	 * @param   integer   $displayType  The type
+	 *
+	 * @return  string
+	 *
+	 * @since   3.7.0
```
Comment (PhilETaylor) [inline plugins/content/fields/fields.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__

File: plugins/content/fields/fields.php [hunk 5]
```diff
@@ -165,4 +171,175 @@ private function prepare($string, $context, $item)
...
+	}
+
+	/**
+	 * Prepares a tag item to be ready for com_fields.
+	 *
+	 * @param   stdClass  $item  The item
+	 *
+	 * @return  object
+	 *
+	 * @since   3.8.4
```
Comment (PhilETaylor) [inline plugins/content/fields/fields.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Comment by piotr-cz** [comment]
@fancyFranci done in https://github.com/joomla/joomla-cms/pull/14875/commits/8287e8c3b6290a75116376175e09be36c4eb5b31

File: plugins/content/fields/fields.php [hunk 6]
```diff
@@ -9,6 +9,12 @@
 
 defined('_JEXEC') or die();
 
+use Joomla\Registry\Registry;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Language\Multilanguage;
```
Comment (Quy) [inline plugins/content/fields/fields.php]:
```suggestion
use Joomla\CMS\Factory;
use Joomla\CMS\Language\Multilanguage;
use Joomla\Registry\Registry;
```

Comment (piotr-cz) [reply plugins/content/fields/fields.php]: Why so?

It's the same order as was in [plugins/system/fields/fields.php](https://github.com/joomla/joomla-cms/pull/14875/files#diff-42d0b38b43b879586c6d4cc04e0434d18bc9ddb1bb251c6e06cf5890aab4a74aL13-L15)

Comment (PhilETaylor) [reply plugins/content/fields/fields.php]: Because A comes before Z :)

Comment (PhilETaylor) [reply plugins/content/fields/fields.php]: "All namespace imports must be alphabetically ordered"

https://developer.joomla.org/coding-standards/php-code.html

Comment (piotr-cz) [reply plugins/content/fields/fields.php]: Okay, thanks

Comment (alikon) [reply plugins/content/fields/fields.php]: alpha order

**Comment by laoneo** [comment]
Can you rebase it to 4.2. So we can test it again and get it in.

**Comment by laoneo** [comment]
Thanks for your contribution with this pr to make Joomla better. This pr is not anymore suitable for Joomla 3. Can you rebase it to the latest Joomla 4 branch so we can reevaluate it again? In the meantime we are closing it, when done, don't hesitate to reopen again.

**Comment by piotr-cz** [comment]
Sorry, it's taken far too long (5 years) and I'm on different stack now.
I hope somebody will take over this and prepare PR for Joomla 4

**Pull request closed without merging**